lib/samples/snap_geometry_edits/snap_geometry_edits.dart

```diff
@@ -0,0 +1,679 @@
+    _geometryEditor.startWithGeometryType(geometryType);
+  }
+
+  void stopAndSave() {
+    // Get the geometry from the geometry editor.
+    final geometry = _geometryEditor.stop();
+
+    if (geometry != null) {
+      if (_selectedGraphic != null) {
+        // If there was a selected graphic being edited, update it.
+        _selectedGraphic!.geometry = geometry;
+        _selectedGraphic!.isVisible = true;
+        // Reset the selected graphic to null.
+        setState(() => _selectedGraphic = null);
+      } else {
+        // If there was no existing graphic, create a new one and add to the graphics overlay.
+        final graphic = Graphic(geometry: geometry);
+        // Apply a symbol to the graphic from the geometry editor style depending on the geometry type.
+        final geometryType = geometry.geometryType;
+        if (geometryType == GeometryType.point ||
+            geometryType == GeometryType.multipoint) {
+          graphic.symbol = _geometryEditorStyle.vertexSymbol;
+        } else if (geometryType == GeometryType.polyline) {
+          graphic.symbol = _geometryEditorStyle.lineSymbol;
+        } else if (geometryType == GeometryType.polygon) {
+          graphic.symbol = _geometryEditorStyle.fillSymbol;
+        }
+        _graphicsOverlay.graphics.add(graphic);
+      }
+    }
+
+    // Reset the selected geometry type to null.
+    setState(() => _selectedGeometryType = null);
+  }
+
+  void stopAndDiscardEdits() {
+    // Stop the geometry editor. No need to capture the geometry as we are discarding.
+    _geometryEditor.stop();
+    if (_selectedGraphic != null) {
+      // If editing a previously existing geometry, reset the selectedGraphic.
+      _selectedGraphic!.isVisible = true;
+      setState(() => _selectedGraphic = null);
+    }
+    // Reset the selected geometry type.
+    setState(() => _selectedGeometryType = null);
+  }
+
+  Widget buildBottomMenu() {
+    return Row(
+      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
+      children: [
+        // A drop down button for selecting geometry type.
+        DropdownButton(
+          alignment: Alignment.center,
+          hint: const Text(
+            'Geometry Type',
+            style: TextStyle(
+              color: Colors.deepPurple,
+            ),
+          ),
+          icon: const Icon(Icons.arrow_drop_down),
+          iconEnabledColor: Colors.deepPurple,
+          iconDisabledColor: Colors.grey,
+          style: const TextStyle(color: Colors.deepPurple),
+          value: _selectedGeometryType,
+          items: _geometryTypeMenuItems,
+          // If the geometry editor is already started then we don't enable editing with another geometry type.
+          onChanged: _geometryEditorIsStarted
+              ? null
+              : (geometryType) => geometryType == null
+                  ? null
+                  : startEditingWithGeometryType(geometryType as GeometryType),
```

Comment:
`geometryType as GeometryType` should be `geometryType!` since all we're doing is getting rid of the nullability.

But I don't really like the `? null` on line 314. We're not really returning a value here. I think what we're trying to express is just
```dart
(geometryType) => if (geometryType != null) startEditingWithGeometryType(geometryType!),
```

Reply:
I'd tried simpler statements here in the other sample but couldn't get there. The UI dropdown doesn't properly get disabled. I'll have another go at simplifying.

Reply:
The actual callback needs to be null for it to be properly disabled, (i.e. greyed out and not able to interact with, as opposed to just can't select another option which is a bad user experience). So if I check `_geometryEditorIsStarted` I can properly set the DropDownButton to null if it is. Note that then the `geometryType` is an `Object?` hence the typecast there.

I've made a tweak to the existing code. The other option would be the below but the UI is a bit weird. I could style the individual items to look more un-selectable and slightly improve things.

```
          onChanged: (geometryType) {
            if (!_geometryEditorIsStarted && geometryType != null) {
              startEditingWithGeometryType(geometryType);
            }
          },
```

Reply:
Ah ok, I think I see what's going on. This is the first time we're using `null` as the value of `onChanged` to disable the widget. And that's correct, so don't change that.

Unfortunately that introduces a small oddity -- because of that `null`, the compiler infers that the widget is a `DropdownButton<Object>` rather than a `DropdownButton<GeometryType>`. The compiler can't be any more specific about the `null`, so it falls back to `<Object>`. And because of that, you end up with `geometryType` being an `Object?` rather than a `GeometryType`.

I'm not sure the best way to address that -- probably if you changed `DropdownButton` to `DropdownButton<GeometryType>`, that would force the issue and everything would go back to normal. But I don't like that as a general rule. Maybe if you change `? (geometryType) {` ... to ... `? (GeometryType? geometryType) {`, that might give the compiler what it needs to narrow down the type.

Or maybe there's a way to cast the `null` to a `ValueChanged<GeometryType?>?`. I wonder if you just do `null as ValueChanged<GeometryType?>?`.

Reply:
Interesting! I've gone with `(GeometryType? geometryType)`.  `DropdownButton<GeometryType>` does also work.